File: src/jsMain/kotlin/kingdom/lite/fresh/KingdomCore.kt
package kingdom.lite.fresh

import js.objects.JsPlainObject
import kotlinx.serialization.Serializable

/**
 * Fresh start - Kingdom data model based on the simplified rules
 */
@Serializable
data class Kingdom(
    val id: String,
    val name: String = "New Kingdom",
    val level: Int = 1,
    val xp: Int = 0,
    val currentTurn: Int = 1,
    val gold: Int = 0,              // Currency
    val unrest: Int = 0,            // Negative status
    val fame: Int = 0,              // Reputation
    val resources: Resources = Resources(),
    val settlements: List<Settlement> = emptyList(),
    val activeEvents: List<String> = emptyList(), // Event IDs
    val modifiers: List<Modifier> = emptyList()
)

@Serializable
data class Resources(
    val food: Int = 0,
    val lumber: Int = 0,
    val ore: Int = 0,
    val stone: Int = 0
)

@Serializable
data class Settlement(
    val id: String,
    val name: String,
    val level: Int = 1,
    val structures: List<String> = emptyList() // Structure IDs
)

@Serializable
data class Modifier(
    val source: String,
    val type: String,
    val value: Int,
    val duration: String = "permanent" // permanent, turns:X, until:condition
)

/**
 * Raw data types for JSON parsing - matching actual JSON schema
 */

// Structure JSON schema
@JsPlainObject
external interface RawStructure {
    val id: String
    val name: String
    val type: String
    val category: String
    val tier: Int
    val effect: String?
    val earnIncomeLevel: String?
    val bonus: Int?
    val skills: Array<String>?
    val construction: RawConstruction?
    val traits: Array<String>
    val special: String?
    val upgradeFrom: String?
}

@JsPlainObject
external interface RawConstruction {
    val resources: RawResourceCost?
}

@JsPlainObject
external interface RawResourceCost {
    val lumber: Int?
    val stone: Int?
    val ore: Int?
    val food: Int?
}

// Player Action JSON schema
@JsPlainObject
external interface RawPlayerAction {
    val id: String
    val name: String
    val category: String
    val description: String
    val skills: Array<RawSkillOption>?
    val effects: RawActionEffects?
    val special: String?
}

@JsPlainObject
external interface RawSkillOption {
    val skill: String
    val description: String
}

@JsPlainObject
external interface RawActionEffects {
    val criticalSuccess: RawActionResult?
    val success: RawActionResult?
    val failure: RawActionResult?
    val criticalFailure: RawActionResult?
}

@JsPlainObject
external interface RawActionResult {
    val description: String
    val modifiers: dynamic  // Complex object, using dynamic for now
}

// Event JSON schema
@JsPlainObject
external interface RawEvent {
    val id: String
    val name: String
    val description: String
    val traits: Array<String>
    val location: String?
    val modifier: Int
    val resolution: String?
    val resolvedOn: Array<String>?
    val stages: Array<RawEventStage>?
    val special: String?
}

@JsPlainObject
external interface RawEventStage {
    val skills: Array<String>?
    val criticalSuccess: RawEventOutcome?
    val success: RawEventOutcome?
    val failure: RawEventOutcome?
    val criticalFailure: RawEventOutcome?
}

@JsPlainObject
external interface RawEventOutcome {
    val msg: String
    val modifiers: Array<RawEventModifier>?
}

@JsPlainObject
external interface RawEventModifier {
    val type: String
    val name: String
    val value: Int
    val selector: String
    val enabled: Boolean
    val turns: Int?
}

// Incident JSON schema
@JsPlainObject
external interface RawIncident {
    val id: String
    val name: String
    val tier: String
    val description: String
    val percentileMin: Int
    val percentileMax: Int
    val skillOptions: Array<RawIncidentSkillOption>
}

@JsPlainObject
external interface RawIncidentSkillOption {
    val skill: String
    val description: String
    val successEffect: String
    val failureEffect: String
    val criticalFailureExtra: String?
}

File: src-ts/core/KingdomCore.ts
// Auto-converted from KingdomCore.kt
// TODO: Review and fix TypeScript-specific issues


// TODO: Review import - import js.objects.JsPlainObject


/**
 * Fresh start - Kingdom data model based on the simplified rules
 */

export interface Kingdom {
  id: string;
  name: string;
  level: number;
  xp: number;
  currentTurn: number;
  gold: number;
  unrest: number;
  fame: number;
  resources: Resources;
},
    const settlements: Array<Settlement> = emptyList();
    const activeEvents: Array<string> = emptyList(), // Event IDs
    const modifiers: Array<Modifier> = emptyList() }
export interface Resources {
  food: number;
  lumber: number;
  ore: number;
  stone: number;
}


export interface Settlement {
  id: string;
  name: string;
  level: number;
  structures: Array<string>;
} // Structure IDs
)


export interface Modifier {
  source: string;
  type: string;
  value: number;
  duration: string;
  turns: X;
  until: condition;
}

/**
 * Raw data types for JSON parsing - matching actual JSON schema
 */

// Structure JSON schema

declare interface RawStructure {
    const id: string
    const name: string
    const type: string
    const category: string
    const tier: number
    const effect: string | null
    const earnIncomeLevel: string | null
    const bonus: number | null
    const skills: Array<string> | null
    const construction: RawConstruction | null
    const traits: Array<string>
    const special: string | null
    const upgradeFrom: string | null
}


declare interface RawConstruction {
    const resources: RawResourceCost | null
}


declare interface RawResourceCost {
    const lumber: number | null
    const stone: number | null
    const ore: number | null
    const food: number | null
}

// Player Action JSON schema

declare interface RawPlayerAction {
    const id: string
    const name: string
    const category: string
    const description: string
    const skills: Array<RawSkillOption> | null
    const effects: RawActionEffects | null
    const special: string | null
}


declare interface RawSkillOption {
    const skill: string
    const description: string
}


declare interface RawActionEffects {
    const criticalSuccess: RawActionResult | null
    const success: RawActionResult | null
    const failure: RawActionResult | null
    const criticalFailure: RawActionResult | null
}


declare interface RawActionResult {
    const description: string
    const modifiers: any  // Complex object, using dynamic for now
}

// Event JSON schema

declare interface RawEvent {
    const id: string
    const name: string
    const description: string
    const traits: Array<string>
    const location: string | null
    const modifier: number
    const resolution: string | null
    const resolvedOn: Array<string> | null
    const stages: Array<RawEventStage> | null
    const special: string | null
}


declare interface RawEventStage {
    const skills: Array<string> | null
    const criticalSuccess: RawEventOutcome | null
    const success: RawEventOutcome | null
    const failure: RawEventOutcome | null
    const criticalFailure: RawEventOutcome | null
}


declare interface RawEventOutcome {
    const msg: string
    const modifiers: Array<RawEventModifier> | null
}


declare interface RawEventModifier {
    const type: string
    const name: string
    const value: number
    const selector: string
    const enabled: boolean
    const turns: number | null
}

// Incident JSON schema

declare interface RawIncident {
    const id: string
    const name: string
    const tier: string
    const description: string
    const percentileMin: number
    const percentileMax: number
    const skillOptions: Array<RawIncidentSkillOption>
}


declare interface RawIncidentSkillOption {
    const skill: string
    const description: string
    const successEffect: string
    const failureEffect: string
    const criticalFailureExtra: string | null
}
